pkg/sysfs/sst_if.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2021 Intel Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sysfs
+
+import (
+	"fmt"
+	"os"
+	"syscall"
+	"unsafe"
+)
+
+// cpuMap holds the logical to punit cpu mapping table
+var cpuMap = make(map[ID]ID)
+
+// punitCPU returns the PUNIT CPU id corresponding a given Linux logical CPU
+func punitCPU(cpu ID) (ID, error) {
+	if id, ok := cpuMap[cpu]; ok {
+		return id, nil
+	}
+
+	id, err := getCPUMapping(cpu)
+	if err == nil {
+		cpuMap[cpu] = id
+	}
+	return id, err
+}
+
+// isstIoctl is a helper for executing ioctls on the linux isst_if device driver
+func isstIoctl(ioctl uintptr, req uintptr) error {
+	f, err := os.Open(isstDevPath)
+	if err != nil {
+		return fmt.Errorf("failed to open isst device %q: %v", isstDevPath, err)
+	}
+	defer f.Close()
+
+	if _, _, errno := syscall.Syscall(syscall.SYS_IOCTL, uintptr(f.Fd()), ioctl, req); errno != 0 {
+		return errno
+	}
+
+	return nil
+}
+
+// getCPUMapping gets mapping of Linux logical CPU numbers to (package-specific)
+// PUNIT CPU number for one cpu
+func getCPUMapping(cpu ID) (ID, error) {
+	req := isstIfCPUMaps{
+		Cmd_count: 1,
+		Cpu_map: [1]isstIfCPUMap{
+			{Logical_cpu: uint32(cpu)},
+		},
+	}
+
+	if err := isstIoctl(ISST_IF_GET_PHY_ID, uintptr(unsafe.Pointer(&req))); err != nil {
+		return -1, fmt.Errorf("failed to get CPU mapping for cpu %d: %v", cpu, err)
+	}
+
+	return ID(req.Cpu_map[0].Physical_cpu), nil
+}
```

Comment:
nitpick: isn't this package-specific `PUNIT CPU number` the same as already dug out and published by the kernel as `/sys/devices/system/cpu/cpu[0-9]*/topology/core_id` ? If it is the same and we seem to start needing it, wouldn't it be better to add `sysfs.CPU.CoreID() ID` and add a one-liner to pkg/sysfs to read it out during CPU discovery ?

Reply:
That is a good question I pondered myself at some point. I don't think it necessarily is. This punit core id solving is done this way in the `intel-speed-select` tool in the Linux kernel sources so I think they know better and it's better to follow their footsteps. I'd leave this now as is. I'll ping the kernel developers and see what they say. If they confirm your suspicion then we can change this (and I'll possibly able to write a kernel patch 😄 ) WDYT?

Reply:
Sure. That's fine. This is not really a concern, more like wondering...